odl/space/cu_ntuples.py

```diff
@@ -847,36 +976,271 @@ def _norm_default(x):
+                 copy_to_gpu=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        vector : array-like, one-dim.
+            Weighting vector of the inner product
+        exponent : positive float
+            Exponent of the norm. For values other than 2.0, the inner
+            product is not defined.
+            If `matrix` is a sparse matrix, only 1.0, 2.0 and `inf`
+            are allowed.
+        dist_using_inner : bool, optional
+            Calculate `dist` using the formula
+
+            norm(x-y)**2 = norm(x)**2 + norm(y)**2 - 2*inner(x, y).real
+
+            This avoids the creation of new arrays and is thus faster
+            for large arrays. On the downside, it will not evaluate to
+            exactly zero for equal (but not identical) `x` and `y`.
+
+            Can only be used if `exponent` is 2.0.
+        copy_to_gpu : bool
+            If `True`, the weights are stored as `CudaFn.Vector`,
+            which consumes GPU memory but results in faster
+            evaluation. If `False`, weights are stored as a NumPy
+            array on the main memory.
+        """
+        # TODO: remove dist_using_inner in favor of a native dist
+        # implementation
+        super().__init__(exponent=exponent, dist_using_inner=dist_using_inner)
+        if not isinstance(vector, CudaFn.Vector):
+            self._vector = np.asarray(vector)
+        else:
+            self._vector = vector
+        if self._vector.dtype == object:
+            raise ValueError('invalid vector {}.'.format(vector))
+        elif self._vector.ndim != 1:
+            raise ValueError('vector {} is {}-dimensional instead of '
+                             '1-dimensional.'
+                             ''.format(vector, self._vector.ndim))
+        if copy_to_gpu and not isinstance(self._vector, CudaFn.Vector):
+            self._vector = CudaFn(self._vector.size,
+                                  self._vector.dtype).element(self._vector)
+
+    @property
+    def vector(self):
+        """Weighting vector of this inner product."""
+        return self._vector
+
+    def vector_is_valid(self):
+        """Test if the vector is a valid weight, i.e. positive.
+
+        Note
+        ----
+        This operation copies the vector to the CPU memory and uses
+        `numpy.all`, which can be very time-consuming in total.
+        """
+        return np.all(self.vector > 0)
+
+    def __eq__(self, other):
+        """`inner.__eq__(other) <==> inner == other`.
+
+        Returns
+        -------
+        equals : bool
+            `True` if `other` is a `CudaFnVectorWeighting` instance
+            with **identical** vector, `False` otherwise.
+
+        See also
+        --------
+        equiv : test for equivalent inner products
+        """
+        if other is self:
+            return True
+
+        return (isinstance(other, CudaFnVectorWeighting) and
+                self.vector is other.vector and
```

Comment:
Err, no. The vector has the same size as the data (potentially GB), and even "quite fast" is not sufficient since this check is done in *every* space comparison, which in turn is done in each `vec in space` check. This check must be very fast, not just quite fast.